verification/trustpolicy/trustpolicy.go

```diff
@@ -0,0 +1,486 @@
+			identityValue := identity[i+1:]
+
+			// notation natively supports x509.subject identities only
+			if identityPrefix == slice.X509Subject {
+				dn, err := pkix.ParseDistinguishedName(identityValue)
+				if err != nil {
+					return err
+				}
+				parsedDNs = append(parsedDNs, parsedDN{RawString: identity, ParsedMap: dn})
+			}
+		}
+	}
+
+	// Verify there are no overlapping DNs
+	if err := validateOverlappingDNs(statement.Name, parsedDNs); err != nil {
+		return err
+	}
+
+	// No error
+	return nil
+}
+
+// validateRegistryScopes validates if the policy document is following the
+// Notary V2 spec rules for registry scopes
+func validateRegistryScopes(policyDoc *Document) error {
+	registryScopeCount := make(map[string]int)
+
+	for _, statement := range policyDoc.TrustPolicies {
+		// Verify registry scopes are valid
+		if len(statement.RegistryScopes) == 0 {
+			return fmt.Errorf("trust policy statement %q has zero registry scopes, it must specify registry scopes with at least one value", statement.Name)
+		}
+		if len(statement.RegistryScopes) > 1 && slice.Contains(slice.Wildcard, statement.RegistryScopes) {
+			return fmt.Errorf("trust policy statement %q uses wildcard registry scope '*', a wildcard scope cannot be used in conjunction with other scope values", statement.Name)
+		}
+		for _, scope := range statement.RegistryScopes {
+			if scope != slice.Wildcard {
+				if err := validateRegistryScopeFormat(scope); err != nil {
+					return err
+				}
+			}
+			registryScopeCount[scope]++
+		}
+	}
+
+	// Verify one policy statement per registry scope
+	for key := range registryScopeCount {
+		if registryScopeCount[key] > 1 {
+			return fmt.Errorf("registry scope %q is present in multiple trust policy statements, one registry scope value can only be associated with one statement", key)
+		}
+	}
+
+	// No error
+	return nil
+}
+
+func validateOverlappingDNs(policyName string, parsedDNs []parsedDN) error {
+	for i, dn1 := range parsedDNs {
+		for j, dn2 := range parsedDNs {
+			if i != j && pkix.IsSubsetDN(dn1.ParsedMap, dn2.ParsedMap) {
+				return fmt.Errorf("trust policy statement %q has overlapping x509 trustedIdentities, %q overlaps with %q", policyName, dn1.RawString, dn2.RawString)
+			}
+		}
+	}
+
+	return nil
+}
+
+// isValidTrustStoreType returns true if the given string is a valid
+// truststore.Type, otherwise false.
+func isValidTrustStoreType(s string) bool {
+	for _, p := range truststore.Types {
+		if s == string(p) {
+			return true
+		}
+	}
+	return false
+}
+
+// GetApplicableTrustPolicy returns a pointer to the deep copied TrustPolicy
+// statement that applies to the given registry URI. If no applicable trust
+// policy is found, returns an error
+// see https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#selecting-a-trust-policy-based-on-artifact-uri
+func GetApplicableTrustPolicy(trustPolicyDoc *Document, artifactReference string) (*TrustPolicy, error) {
```

Comment:
Should we make this method of type Document
```suggestion
func (policyDoc *Document) GetApplicableTrustPolicy(artifactReference string) (*TrustPolicy, error) {
```